### PR TITLE
[clang] Diagnose functions with too many parameters

### LINKED ISSUE #35741: front end assertion: function has too many parameters
|  |  |
| --- | --- |
| Bugzilla Link | [36393](https://llvm.org/bz36393) |
| Version | trunk |
| OS | All |
| Attachments | [Reproducer](https://user-images.githubusercontent.com/60944935/143757044-634d8d70-5c49-4071-bd88-6d667bcc150c.gz) |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 

Attaching a test case from the gcc testsuite.  Compile with clang -O0 -S.  I get:
```
clang-6.0: /home/ruttenberg/reservoir/px/llvm/clang/lib/AST/Type.cpp:2729:
clang::FunctionProtoType::FunctionProtoType(clang::QualType, llvm::ArrayRef<clang::QualType>, clang::QualType, const clang::FunctionProtoType::ExtProtoInfo&):
Assertion `NumParams == params.size() && "function has too many parameters"' failed.

#0 0x00007ff9b766f86a llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/ruttenberg/reservoir/px/llvm/llvm/lib/Support/Unix/Signals.inc:402:0
#1 0x00007ff9b766d6fe llvm::sys::RunSignalHandlers() /home/ruttenberg/reservoir/px/llvm/llvm/lib/Support/Signals.cpp:50:0
#2 0x00007ff9b766d872 SignalHandler(int) /home/ruttenberg/reservoir/px/llvm/llvm/lib/Support/Unix/Signals.inc:242:0
#3 0x00007ff9b50bc4b0 (/lib/x86_64-linux-gnu/libc.so.6+0x354b0)
#4 0x00007ff9b50bc428 gsignal /build/glibc-Cl5G7W/glibc-2.23/signal/../sysdeps/unix/sysv/linux/raise.c:54:0
#5 0x00007ff9b50be02a abort /build/glibc-Cl5G7W/glibc-2.23/stdlib/abort.c:91:0
#6 0x00007ff9b50b4bd7 __assert_fail_base /build/glibc-Cl5G7W/glibc-2.23/assert/assert.c:92:0
#7 0x00007ff9b50b4c82 (/lib/x86_64-linux-gnu/libc.so.6+0x2dc82)
#8 0x00007ff9b20a1ce0 clang::FunctionProtoType::canThrow(clang::ASTContext const&) const /home/ruttenberg/reservoir/px/llvm/clang/lib/AST/Type.cpp:2860:0
#9 0x00007ff9b1eb638a clang::ASTContext::getFunctionTypeInternal(clang::QualType, llvm::ArrayRef<clang::QualType>, clang::FunctionProtoType::ExtProtoInfo const&, bool) const /home/ruttenberg/reservoir/px/llvm/clang/lib/AST/ASTContext.cpp:3610:0
#10 0x00007ff9b04c75db GetFullTypeForDeclarator((anonymous namespace)::TypeProcessingState&, clang::QualType, clang::TypeSourceInfo*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Sema/SemaType.cpp:4643:0
#11 0x00007ff9b04c94d3 ~TypeProcessingState /home/ruttenberg/reservoir/px/llvm/clang/lib/Sema/SemaType.cpp:229:0
#12 0x00007ff9b04c94d3 clang::Sema::GetTypeForDeclarator(clang::Declarator&, clang::Scope*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Sema/SemaType.cpp:4953:0
#13 0x00007ff9b00ea83e clang::TypeSourceInfo::getType() const /home/ruttenberg/reservoir/px/llvm/clang/include/clang/AST/Decl.h:97:0
#14 0x00007ff9b00ea83e clang::Sema::HandleDeclarator(clang::Scope*, clang::Declarator&, llvm::MutableArrayRef<clang::TemplateParameterList*>) /home/ruttenberg/reservoir/px/llvm/clang/lib/Sema/SemaDecl.cpp:5354:0
#15 0x00007ff9b00edd6f clang::Sema::ActOnStartOfFunctionDef(clang::Scope*, clang::Declarator&, llvm::MutableArrayRef<clang::TemplateParameterList*>, clang::Sema::SkipBodyInfo*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Sema/SemaDecl.cpp:11833:0
#16 0x00007ff9b0a36421 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:1147:0
#17 0x00007ff9b09b4db6 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, unsigned int, clang::SourceLocation*, clang::Parser::ForRangeInit*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/ParseDecl.cpp:1961:0
#18 0x00007ff9b0a3276a clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:987:0
#19 0x00007ff9b0a32e71 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) [clone .part.161] [clone .constprop.184] /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:1003:0
#20 0x00007ff9b0a32ecf clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:1005:0
#21 0x00007ff9b0a3777d clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:853:0
#22 0x00007ff9b0a3878c clang::AttributePool::~AttributePool() /home/ruttenberg/reservoir/px/llvm/clang/include/clang/Sema/AttributeList.h:643:0
#23 0x00007ff9b0a3878c clang::ParsedAttributes::~ParsedAttributes() /home/ruttenberg/reservoir/px/llvm/clang/include/clang/Sema/AttributeList.h:734:0
#24 0x00007ff9b0a3878c clang::Parser::ParsedAttributesWithRange::~ParsedAttributesWithRange() /home/ruttenberg/reservoir/px/llvm/clang/include/clang/Parse/Parser.h:1304:0
#25 0x00007ff9b0a3878c clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:606:0
#26 0x00007ff9b0a38a10 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/Parser.cpp:536:0
#27 0x00007ff9b099720d clang::ParseAST(clang::Sema&, bool, bool) /home/ruttenberg/reservoir/px/llvm/clang/lib/Parse/ParseAST.cpp:146:0
#28 0x00007ff9b638532f clang::CodeGenAction::ExecuteAction() /home/ruttenberg/reservoir/px/llvm/clang/lib/CodeGen/CodeGenAction.cpp:1032:0
#29 0x00007ff9b5a98c96 clang::FrontendAction::Execute() /home/ruttenberg/reservoir/px/llvm/clang/lib/Frontend/FrontendAction.cpp:897:0
#30 0x00007ff9b5a5e766 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ruttenberg/reservoir/px/llvm/clang/lib/Frontend/CompilerInstance.cpp:992:0
#31 0x00007ff9b57d65f2 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ruttenberg/reservoir/px/llvm/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:252:0
#32 0x000000000040fd58 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ruttenberg/reservoir/px/llvm/clang/tools/driver/cc1_main.cpp:221:0
#33 0x000000000040b3f7 ExecuteCC1Tool /home/ruttenberg/reservoir/px/llvm/clang/tools/driver/driver.cpp:309:0
#34 0x000000000040b3f7 main /home/ruttenberg/reservoir/px/llvm/clang/tools/driver/driver.cpp:388:0
#35 0x00007ff9b50a7830 __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:325:0
#36 0x000000000040c689 _start (/mnt/ssd/ruttenberg/reservoir/px/install/bin/clang-6.0+0x40c689)
Stack dump:
```

### PR DESCRIPTION
**Pull request by Endilll**
This patch adds a parser check when a function declaration or function type declaration (in a function pointer declaration, for example) has too many parameters for `FunctionTypeBits::NumParams` to hold. At the moment of writing it's a 16-bit-wide bit-field, limiting the number of parameters at 65536.

The check is added in the parser loop that goes over comma-separated list of function parameters. This is not the solution Aaron suggested in https://github.com/llvm/llvm-project/issues/35741#issuecomment-1638086571, because it was found out that it's quite hard to recover from this particular error in `GetFullTypeForDeclarator()`. Multiple options were tried, but all of them led to crashes down the line.

I used LLVM Compile Time Tracker to ensure this does not introduce a performance regression. I believe changes are in the noise: https://llvm-compile-time-tracker.com/compare.php?from=de5ea2d122c31e1551654ff506c33df299f351b8&to=424818620766cedb2770e076ee359afeb0cc14ec&stat=instructions:u

Fixes #35741

### CODE REVIEW

File: clang/lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -8025,6 +8025,17 @@ void Parser::ParseParameterDeclarationClause(
         // Consume the keyword.
         ConsumeToken();
       }
+
+      // FunctionTypeBitfields::NumParams can only hold so much.
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseDecl.cpp:8029]:
```suggestion
      // We can only store so many parameters.
```

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -9311,6 +9311,9 @@ def note_constinit_missing_here : Note<
 def err_dimension_expr_not_constant_integer : Error<
   "dimension expression does not evaluate to a constant unsigned int">;
 
+def err_function_parameter_limit_exceeded : Error<
+  "too many function parameters; subsequent parameters are skipped">;
```
Comment (cor3ntin) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:9315]:
```suggestion
  "too many function parameters; subsequent parameters will be ignored">, DefaultFatal;
```

**Review by cor3ntin — APPROVED** [review]
LGTM

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (Endilll) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: I changed the wording, but I don't agree with `DefaultFatal`. We use it very sparingly (only 51 result across all of our diagnostics), and now we're successfully recovering.
CC @AaronBallman

**Comment by cor3ntin** [comment]
@Endilll  you forgot to update the tests - hence the ci failure

**Comment by Endilll** [comment]
> @Endilll you forgot to update the tests - hence the ci failure

Thank you! Should be good to go now

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: If we can successfully recover, then it's better to drop the `DefaultFatal` so users can get additional follow-on diagnostics. `DefaultFatal` is more for "there's no decent recovery and continuing would be a bad idea"

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**